async_cache.go

```diff
@@ -95,3 +103,63 @@ func (ks *keyStatus) purge() {
 		}
 	}()
 }
+
+//Init NewAsyncCache
+// set cache_expiration to 0 , to use default EXPIRATION_TIME --> 30 Minutes
+func NewAsyncCache(prefix_len int, purge_time time.Duration, cache_expiration time.Duration) *AsyncCache {
+	if cache_expiration == 0 {
+		cache_expiration = EXPIRATION_TIME
+	}
+	return &AsyncCache{
+		fetcher:   NewFetcher(prefix_len),
+		keystatus: NewKeyStatus(purge_time),
+		gCache:    New(cache_expiration, purge_time),
+	}
+}
+
+func (ac *AsyncCache) AsyncGet(key string) (interface{}, status, error) {
+	//Fetching from cache
+	cache_status := STATUS_INPROCESS
+	data, ok := ac.gCache.Get(key)
+	if ok {
+		return data, STATUS_DONE, nil
+	} else if !ok && ac.keystatus.Get(key) == STATUS_INPROCESS {
+		//data not present in cache and status INPROCESS
+		return data, cache_status, nil
+	} else if ac.keystatus.Get(key) == STATUS_INTERNAL_ERROR {
+		//returning error for internal error
+		return data, STATUS_INTERNAL_ERROR, errors.New("error : internal error from data source")
+	} else if !ok && ac.keystatus.Get(key) != STATUS_INPROCESS {
+		//New Call for the key, updating KeyStatus for key to Status INPROCESS
+		ac.keystatus.Set(key, cache_status)
+		//asyncCall to DB/dataSource
+		go asyncUpdate(ac, key)
+	}
+	//returning data and cache/keystatus
+	return data, cache_status, nil
+}
+
+func asyncUpdate(ac *AsyncCache, key string) {
+	done := make(chan bool, 1)
+	go func(done_status chan bool) {
+		dataStatus := STATUS_INPROCESS
+		fetchedData, err := ac.fetcher.Execute(key)
+		// fetching and returning data
+		if err != nil {
+			// Response Error from DB/Fetcher error
+			dataStatus = STATUS_INTERNAL_ERROR
+			ac.keystatus.Set(key, dataStatus)
+			done <- true
+			return
+		}
+		ac.gCache.Set(key, fetchedData, ac.keystatus.purgeTime)
```

Comment:
cache data should get set with expiry time not with purge time interval, please pass expiry time here